src/telemetry/mod.rs

```diff
@@ -1,9 +1,41 @@
 pub mod metrics {
-    use metrics::describe_counter;
+    use axum::http::Version;
     use metrics::Unit;
+    use metrics::{describe_counter, KeyName};
 
     pub const REQUESTS_RECEIVED: &str = "requests.received";
 
+    /// Metric that records the version of HTTP protocol used for a particular request.
+    #[cfg(feature = "web-app")]
+    pub struct RequestProtocolVersion(Version);
+
+    #[cfg(feature = "web-app")]
+    impl From<Version> for RequestProtocolVersion {
+        fn from(v: Version) -> Self {
+            RequestProtocolVersion(v)
+        }
+    }
+
+    #[cfg(feature = "web-app")]
+    impl From<RequestProtocolVersion> for KeyName {
+        fn from(v: RequestProtocolVersion) -> Self {
+            const HTTP11: &str = "request.protocol.HTTP/1.1";
+            const HTTP2: &str = "request.protocol.HTTP/2.0";
+            const HTTP3: &str = "request.protocol.HTTP/3.0";
+            const UNKNOWN: &str = "request.protocol.HTTP/UNKNOWN";
+
+            KeyName::from_const_str(if v.0 == Version::HTTP_11 {
+                HTTP11
+            } else if v.0 == Version::HTTP_2 {
+                HTTP2
+            } else if v.0 == Version::HTTP_3 {
+                HTTP3
+            } else {
+                UNKNOWN
+            })
```

Comment:
would a match statement be an improvement here?
```suggestion
            KeyName::from_const_str(match v.0 {
                Version::HTTP_11 => HTTP11,
                Version::HTTP_2 => HTTP2,
                Version::HTTP_3 => HTTP3,
                _ => UNKNOWN,
            })
```

Reply:
I don't think you can do that because `Version` is not an enum and these values are just constants :(

Reply:
Try it.  Also, why is this doing a string lookup rather than taking an enum?

Reply:
This is the definition of `Version` struct - it is not an enum and `Http` is not accessible from it

```rust
/// Represents a version of the HTTP spec.
#[derive(PartialEq, PartialOrd, Copy, Clone, Eq, Ord, Hash)]
pub struct Version(Http);

impl Version {
    /// `HTTP/0.9`
    pub const HTTP_09: Version = Version(Http::Http09);

    /// `HTTP/1.0`
    pub const HTTP_10: Version = Version(Http::Http10);

    /// `HTTP/1.1`
    pub const HTTP_11: Version = Version(Http::Http11);

    /// `HTTP/2.0`
    pub const HTTP_2: Version = Version(Http::H2);

    /// `HTTP/3.0`
    pub const HTTP_3: Version = Version(Http::H3);
}

```

Reply:
The match should work: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=503992d87137d897f95361095e3ac546

Reply:
nice! changed it to the proposed match